docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,628 @@
+                "torch.nn.parameter.Parameter" | and subclasses of torch.Tensor
+
+
+Special note on torch.nn.ModuleList and torch.nn.ModuleDict
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+Although torch.nn.ModuleList and torch.nn.ModuleDict are defined as a list and dictionary in Python,
+in TorchScript, ``torch.nn.ModuleList`` and ``torch.nn.ModuleDict`` behave more like a tuple.
+
+* Within the execution of TorchScript, instances of ``torch.nn.ModuleList``  or ``torch.nn.ModuleDict`` are immutable.
+* Iterating over ``torch.nn.ModuleList`` or ``torch.nn.ModuleDict`` is completely unrolled so that elements of ``torch.nn.ModuleList`` or keys of ``torch.nn.ModuleDict`` can be of different subclasses of ``torch.nn.Module``.
+
+Example:
+
+.. testcode::
+
+    import torch
+
+    @torch.jit.script
+    class A():
+        def __init__(self):
+            self.x = torch.rand(3)
+
+        def f(self, y: torch.device):
+            return self.x.to(device=y)
+
+    def g():
+        a = A()
+        return a.f(torch.device("cpu"))
+
+    script_g = torch.jit.script(g)
+    print(script_g.graph)
+
+Custom Class
+^^^^^^^^^^^^
+
+Unlike built-in classes, semantics of custom classes are user-defined and the entire class definition must be compilable into TorchScript IR and subject to TorchScript type-checking rules.
+
+::
+
+    TSClassDef := [ "@torch.jit.script" ]
+                "class" ClassName [ "(object)" ]  ":"
+                        MethodDefinition |
+                    [ "@torch.jit.ignore" ] | [ "@torch.jit.unused" ]
+                        MethodDefinition
+
+where
+
+* Classes must be new-style classes (note that Python 3 supports only new-style classes, for Python 2.x new-style class is specified by subclassing from object)
+* Instance data attributes are statically typed, and instance attributes must be declared by assignments inside the ``__init__()`` method
+* Method overloading is not supported (i.e., cannot have multiple methods with the same method name)
+* MethodDefinition must be compilable to TorchScript IR and subject to TorchScript’s type-checking rules, i.e., all methods must be valid TorchScript functions and class attribute definitions are valid TorchScript statements
+* ``torch.jit.ignore`` and ``torch.jit.unused`` can be used to ignore the method or function that is not fully torchscriptable or just needs to be ignored by the compiler
+
+Compared to Python
+""""""""""""""""""
+
+TorchScript custom class is quite limited compared to its Python counterpart. Most notably, TorchScript custom classes
+
+* do not support class attributes
+* do not support subclassing except for subclassing an interface type or object
+* do not support method overloading
+* must initialize all its instance attributes in  ``__init__()``; this is because TorchScript construct a static schema of the class by inferring attribute types in ``__init__()``
```

Comment:
```suggestion
* must initialize all its instance attributes in  ``__init__()``; this is because TorchScript constructs a static schema of the class by inferring attribute types in ``__init__()``
```